File: yonnu_makk_front/lib/models/grossesse.dart
// ─── MODÈLE GROSSESSE ─────────────────────────────────────────────────────────
// Correspond à GrossesseResource du back-end.
class Grossesse {
  final int     id;
  final String? dateDebutGrossesse;      // 'Y-m-d'
  final String? dateAccouchementPrevue;  // 'Y-m-d'
  final String? groupeSanguin;
  final int     nombreGrossessesAnterieures;
  final int     nombreAccouchementsAnterieurs;
  final String? antecedentsObstetricaux;
  final bool    grossesseActive;
  final int?    semainesAmenorrhee;      // semaines_amenorrhee_actuelles
  final int?    joursAvantAccouchement;

  const Grossesse({
    required this.id,
    this.dateDebutGrossesse,
    this.dateAccouchementPrevue,
    this.groupeSanguin,
    this.nombreGrossessesAnterieures = 0,
    this.nombreAccouchementsAnterieurs = 0,
    this.antecedentsObstetricaux,
    this.grossesseActive = true,
    this.semainesAmenorrhee,
    this.joursAvantAccouchement,
  });

  factory Grossesse.fromJson(Map<String, dynamic> json) => Grossesse(
    id:                            json['id'] as int,
    dateDebutGrossesse:            json['date_debut_grossesse'] as String?,
    dateAccouchementPrevue:        json['date_accouchement_prevue'] as String?,
    groupeSanguin:                 json['groupe_sanguin'] as String?,
    nombreGrossessesAnterieures:   (json['nombre_grossesses_anterieures'] as int?) ?? 0,
    nombreAccouchementsAnterieurs: (json['nombre_accouchements_anterieurs'] as int?) ?? 0,
    antecedentsObstetricaux:       json['antecedents_obstetricaux'] as String?,
    grossesseActive:               json['grossesse_active'] as bool? ?? true,
    semainesAmenorrhee:            json['semaines_amenorrhee_actuelles'] as int?,
    joursAvantAccouchement:        json['jours_avant_accouchement'] as int?,
  );

  // Trimestre déduit des semaines d'aménorrhée.
  int get trimestre {
    final sa = semainesAmenorrhee ?? 0;
    if (sa <= 13) return 1;
    if (sa <= 27) return 2;
    return 3;
  }
}

File: app/Models/Grossesse.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Grossesse extends Model
{
    use HasFactory;

    protected $fillable = [
        'femme_id',
        'date_debut_grossesse',
        'date_accouchement_prevue',
        'groupe_sanguin',
        'nombre_grossesses_anterieures',
        'nombre_accouchements_anterieurs',
        'antecedents_obstetricaux',
        'grossesse_active',
    ];

    protected $casts = [
        'date_debut_grossesse'     => 'date',
        'date_accouchement_prevue' => 'date',
        'grossesse_active'         => 'boolean',
    ];

    public function femme(): BelongsTo
    {
        return $this->belongsTo(Femme::class);
    }

    public function suivis(): HasMany
    {
        return $this->hasMany(SuiviGrossesse::class);
    }

    public function mouvements(): HasMany
    {
        return $this->hasMany(MouvementBebe::class);
    }

    public function semainesAmenorrhee(): int
    {
        // Semaines écoulées depuis le début de grossesse (toujours positif).
        // diffInWeeks est signé sous Carbon 3 : on part de la date de début vers maintenant.
        return (int) $this->date_debut_grossesse->diffInWeeks(now());
    }

    public function joursAvantAccouchement(): int
    {
        return (int) now()->diffInDays($this->date_accouchement_prevue, false);
    }
}
